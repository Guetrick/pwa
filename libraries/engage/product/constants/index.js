// MEDIA TYPES
export const MEDIA_TYPE_IMAGE = 'image';
export const MEDIA_TYPE_VIDEO = 'video';

// AVAILABILITY TYPES
export const AVAILABILITY_TYPE_AVAILABLE = 'available';
export const AVAILABILITY_TYPE_LIMITED_AVAILABILITY = 'limitedAvailability';
export const AVAILABILITY_TYPE_NOT_AVAILABLE = 'notAvailable';
export const AVAILABILITY_TYPE_CUSTOM_AVAILABILITY = 'customAvailability';
export const availabilityTypes = [
  AVAILABILITY_TYPE_AVAILABLE,
  AVAILABILITY_TYPE_LIMITED_AVAILABILITY,
  AVAILABILITY_TYPE_NOT_AVAILABLE,
  AVAILABILITY_TYPE_CUSTOM_AVAILABILITY,
];

export const PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP = 'directShip';
export const PRODUCT_FULFILLMENT_METHOD_IN_STORE_PICKUP = 'inStorePickup';
