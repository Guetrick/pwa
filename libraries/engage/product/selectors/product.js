import { createSelector } from 'reselect';
import {
  getProductPropertiesState,
  getProductId,
} from '@shopgate/pwa-common-commerce/product/selectors/product';
import { filterProperties } from './helpers';

/**
 * Creates the selector to get a product's properties from the state.
 * @returns {Function}
 */
export function makeGetProductProperties() {
  return createSelector(
    getProductPropertiesState,
    getProductId,
    (properties, productId) => {
      const entry = properties[productId];

      if (!entry || !entry.properties) {
        return null;
      }

      return filterProperties(entry.properties);
    }
  );
}
