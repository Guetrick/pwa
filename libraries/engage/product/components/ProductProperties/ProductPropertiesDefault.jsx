import React from 'react';
import PropTypes from 'prop-types';
import { SurroundPortals } from '@shopgate/pwa-common/components';
import { PRODUCT_PROPERTIES } from '@shopgate/pwa-common-commerce/product/constants/Portals';
import { ProductPropertiesRows } from './ProductPropertiesRows';
import { container } from './ProductProperties.style';
import connect from './ProductPropertiesDefault.connector';

/**
 * @param {Object} props The component props.
 * @returns {JSX}
 */
function ProductPropertiesDefaultContainer({ properties }) {
  return (
    <SurroundPortals portalName={PRODUCT_PROPERTIES}>
      {(properties !== null && properties.length !== 0) && (
        <div className={container}>
          <table>
            <tbody>
              <ProductPropertiesRows properties={properties} />
            </tbody>
          </table>
        </div>
      )}
    </SurroundPortals>
  );
}

ProductPropertiesDefaultContainer.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()),
};

ProductPropertiesDefaultContainer.defaultProps = {
  properties: null,
};

export const ProductPropertiesDefault = connect(ProductPropertiesDefaultContainer);
