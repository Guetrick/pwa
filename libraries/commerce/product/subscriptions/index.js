import { hex2bin } from '@shopgate/pwa-common/helpers/data';
import getProduct from '../actions/getProduct';
import getProductDescription from '../actions/getProductDescription';
import getProductProperties from '../actions/getProductProperties';
import getProductImages from '../actions/getProductImages';
import getProductShipping from '../actions/getProductShipping';
import getProductVariants from '../actions/getProductVariants';
import getProductOptions from '../actions/getProductOptions';
import {
  productWillEnter$,
  productReceived$,
  cachedProductReceived$,
  productRelationsReceived$,
} from '../streams';
import getProductsById from '../actions/getProductsById';
import { getProductRelationsByHash } from '../selectors/relations';

/**
 * Product subscriptions.
 * @param {Function} subscribe The subscribe function.
 */
function product(subscribe) {
  const processProduct$ = productReceived$.merge(cachedProductReceived$);

  subscribe(productWillEnter$, ({ action, dispatch }) => {
    const { productId } = action.route.params;
    const id = hex2bin(productId);

    dispatch(getProduct(id));
    dispatch(getProductDescription(id));
    dispatch(getProductProperties(id));
    dispatch(getProductImages(id));
    dispatch(getProductShipping(id));
  });

  subscribe(processProduct$, ({ action, dispatch }) => {
    const {
      id,
      flags = {
        hasVariants: false,
        hasOptions: false,
      },
      baseProductId,
    } = action.productData;

    if (baseProductId) {
      dispatch(getProduct(baseProductId));
      dispatch(getProductImages(baseProductId));
    }

    if (flags.hasVariants) {
      dispatch(getProductVariants(id));
    }

    if (flags.hasOptions) {
      dispatch(getProductOptions(id));
    }
  });

  subscribe(productRelationsReceived$, ({ dispatch, getState, action }) => {
    const { hash } = action;
    const productIds = getProductRelationsByHash(hash)(getState());

    dispatch(getProductsById(productIds));
  });
}

export default product;
