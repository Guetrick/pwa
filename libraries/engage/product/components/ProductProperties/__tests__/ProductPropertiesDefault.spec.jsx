import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { getProductProperties } from '@shopgate/pwa-common-commerce/product';
import { ProductPropertiesDefault } from '../ProductPropertiesDefault';

const mockStore = configureStore();
const store = mockStore({});

jest.mock('@shopgate/pwa-common-commerce/product', () => ({
  getProductProperties: jest.fn(),
}));

describe('<ProductPropertiesDefault />', () => {
  it('should not render if no data is available', () => {
    getProductProperties.mockReturnValueOnce(null);
    const wrapper = shallow(<ProductPropertiesDefault store={store} />);
    expect(wrapper.find('table').length).toEqual(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render table with properties', () => {
    getProductProperties.mockReturnValueOnce([
      {
        label: 'test1',
        value: '123',
      },
      {
        label: 'test2',
        value: '456',
      },
    ]);
    const wrapper = mount(<ProductPropertiesDefault store={store} />);
    expect(wrapper.find('table').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });
});
