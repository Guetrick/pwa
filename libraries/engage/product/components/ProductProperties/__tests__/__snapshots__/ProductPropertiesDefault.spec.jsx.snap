// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductPropertiesDefault /> should not render if no data is available 1`] = `
<ProductPropertiesDefaultContainer
  dispatch={[Function]}
  properties={null}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`<ProductPropertiesDefault /> should render table with properties 1`] = `
<Connect(ProductPropertiesDefaultContainer)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ProductPropertiesDefaultContainer
    dispatch={[Function]}
    properties={
      Array [
        Object {
          "label": "test1",
          "value": "123",
        },
        Object {
          "label": "test2",
          "value": "456",
        },
      ]
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <SurroundPortals
      portalName="product.properties"
      portalProps={null}
    >
      <Portal
        name="product.properties.before"
        props={null}
      />
      <Portal
        name="product.properties"
        props={null}
      >
        <div
          className={
            Object {
              "data-css-1x8awm9": "",
            }
          }
        >
          <table>
            <tbody>
              <ProductPropertiesRows
                properties={
                  Array [
                    Object {
                      "label": "test1",
                      "value": "123",
                    },
                    Object {
                      "label": "test2",
                      "value": "456",
                    },
                  ]
                }
              >
                <tr
                  key="test1123"
                >
                  <td
                    className={
                      Object {
                        "data-css-fl9ufm": "",
                      }
                    }
                  >
                    test1
                  </td>
                  <td
                    className={
                      Object {
                        "data-css-fl9ufm": "",
                      }
                    }
                    data-test-id="property: 123"
                  >
                    123
                  </td>
                </tr>
                <tr
                  key="test2456"
                >
                  <td
                    className={
                      Object {
                        "data-css-fl9ufm": "",
                      }
                    }
                  >
                    test2
                  </td>
                  <td
                    className={
                      Object {
                        "data-css-fl9ufm": "",
                      }
                    }
                    data-test-id="property: 456"
                  >
                    456
                  </td>
                </tr>
              </ProductPropertiesRows>
            </tbody>
          </table>
        </div>
      </Portal>
      <Portal
        name="product.properties.after"
        props={null}
      />
    </SurroundPortals>
  </ProductPropertiesDefaultContainer>
</Connect(ProductPropertiesDefaultContainer)>
`;
